lib/constructs/callback-interface.js

```diff
@@ -0,0 +1,233 @@
+            );
+          }
+          this.operation = member;
+          break;
+        case "const":
+          this.constants.set(member.name, new Constant(this.ctx, this, member));
+          break;
+        default:
+          if (!this.ctx.options.suppressErrors) {
+            throw new Error(
+              `Unknown IDL member type "${member.type}" in callback interface ${this.name}`
+            );
+          }
+      }
+    }
+
+    if (this.operation === null) {
+      throw new Error(`Callback interface ${this.name} has no operation`);
+    }
+  }
+
+  addAllProperties() {
+    for (const member of this.constants.values()) {
+      const data = member.generate();
+      this.requires.merge(data.requires);
+    }
+  }
+
+  addStaticProperty(name, body, { configurable = true, enumerable = typeof name === "string", writable = true } = {}) {
+    const descriptor = { configurable, enumerable, writable };
+    this._outputStaticProperties.set(name, { body, descriptor });
+  }
+
+  addProperty() {}
+
+  generateConversion() {
+    const { operation, name } = this;
+    const opName = operation.name;
+
+    const isAsyncInUnion = operation.idlType.union &&
+                           operation.idlType.idlType.some(idlType => idlType.generic === "Promise");
+    const isAsync = isAsyncInUnion || operation.idlType.generic === "Promise";
+
+    const argNames = operation.arguments.map(arg => arg.name);
+    if (operation.arguments.some(arg => arg.optional || arg.variadic)) {
+      throw new Error("Internal error: optional/variadic arguments are not implemented for callback interfaces");
+    }
+
+    this.str += `
+      exports.convert = function convert(value, { context = "The provided value" } = {}) {
+        if (!utils.isObject(value)) {
+          throw new TypeError(\`\${context} is not an object.\`);
+        }
+
+        function callTheUserObjectsOperation(${argNames.join(", ")}) {
+          let thisArg = this;
+          let O = value;
+          let X = O;
+    `;
+
+    if (isAsync) {
+      this.str += `
+          try {
+      `;
+    }
+
+    this.str += `
+            if (typeof O !== "function") {
+              X = O[${utils.stringifyPropertyName(opName)}];
+              if (typeof X !== "function") {
+                throw new TypeError(\`\${context} does not correctly implement ${name}.\`)
+              }
+              thisArg = O;
+            }
+    `;
+
+    // We don't implement all of https://heycam.github.io/webidl/#web-idl-arguments-list-converting since the callers
+    // are assumed to always pass the correct number of arguments and we don't support optional/variadic arguments.
+    for (const argName of argNames) {
+      this.str += `
+            ${argName} = utils.tryWrapperForImpl(${argName});
+      `;
+    }
+
+    const argsToReflectCall = argNames.length > 0 ? `, ${argNames.join(", ")}` : "";
+    this.str += `
+            let callResult = Reflect.call(X, thisArg${argsToReflectCall});
```

Comment:
There is no such thing as `Reflect.call(…)`, so this line will throw a `TypeError`.

I’ve fixed that in <https://github.com/jsdom/webidl2js/pull/172>.